lib/optimizer.py

```diff
@@ -88,24 +88,55 @@ def _WorksBetterOnSomeOtherClip(self, encoding, bitrate, videofile):
           return best_on_this
     return None
 
-  def _EncodingWithOneLessParameter(self, encoding, bitrate, videofile):
+  def _EncodingWithOneLessParameter(self, encoding, bitrate, videofile,
+                                    considered_seen):
     """Find an untried encoder that has one parameter less than this one."""
+    # pylint: disable=R0201
     new_encoder = encoding.encoder.RandomlyRemoveParameter()
     if not new_encoder:
       return None
+    if considered_seen and new_encoder.Hashname() in considered_seen:
+      return None
     new_encoding = new_encoder.Encoding(bitrate, videofile)
     new_encoding.Recover()
     return new_encoding
 
-  def BestUntriedEncoding(self, bitrate, videofile):
-    """Attempts to guess the best untried encoding for this file and rate."""
+  # pylint: disable=unused-argument
+  def _EncodingGoodOnOtherRate(self, encoding, bitrate, videofile,
+                               considered_seen):
+    """Find an untried encoder that is "best" on some other bitrate."""
+
+    if not self.file_set:
+      return None
+    for other_rate in self.file_set.AllRatesForFile(videofile.filename):
+      new_encoder = self.BestEncoding(other_rate, videofile).encoder
+      if considered_seen and new_encoder.Hashname() in considered_seen:
+        continue
+      new_encoding = new_encoder.Encoding(bitrate, videofile)
+      new_encoding.Recover()
+      if not new_encoding.Result():
+        return new_encoding
+    return None
+
+  def BestUntriedEncoding(self, bitrate, videofile, considered_seen=None):
+    """Attempts to guess the best untried encoding for this file and rate.
+
+    Arguments:
+    - bitrate - target bitrate in kbits/sec.
+    - videofile - encoder.Videofile object for the file to be encoded.
+    - considered_seen - list of hashnames for encoders that are not wanted.
```

Comment:
considered_seen is a confusing name in this context; however it does make sense in the calling context. Perhaps ignored_encoder_hashnames is a better name here? Also it's a set, not a list right?